fail2ban/tests/fail2banclienttestcase.py

```diff
@@ -0,0 +1,598 @@
+# emacs: -*- mode: python; py-indent-offset: 4; indent-tabs-mode: t -*-
+# vi: set ft=python sts=4 ts=4 sw=4 noet :
+
+# This file is part of Fail2Ban.
+#
+# Fail2Ban is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# Fail2Ban is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Fail2Ban; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+
+# Fail2Ban developers
+
+__author__ = "Serg Brester"
+__copyright__ = "Copyright (c) 2014- Serg G. Brester (sebres), 2008- Fail2Ban Contributors"
+__license__ = "GPL"
+
+import fileinput
+import os
+import re
+import sys
+import time
+import signal
+import unittest
+
+from functools import wraps
+from threading import Thread
+
+
+from ..client import fail2banclient, fail2banserver, fail2bancmdline
+from ..client.fail2banclient import Fail2banClient, exec_command_line as _exec_client, VisualWait
+from ..client.fail2banserver import Fail2banServer, exec_command_line as _exec_server
+from .. import protocol
+from ..server import server
+from ..server.utils import Utils
+from .utils import LogCaptureTestCase, logSys, with_tmpdir, shutil, logging
+
+
+STOCK_CONF_DIR = "config"
+STOCK = os.path.exists(os.path.join(STOCK_CONF_DIR,'fail2ban.conf'))
+
+CLIENT = "fail2ban-client"
+SERVER = "fail2ban-server"
+BIN = os.path.dirname(Fail2banServer.getServerPath())
+
+MAX_WAITTIME = 30 if not unittest.F2B.fast else 5
+
+##
+# Several wrappers and settings for proper testing:
+#
+
+fail2bancmdline.MAX_WAITTIME = MAX_WAITTIME-1
+
+fail2bancmdline.logSys = \
+fail2banclient.logSys = \
+fail2banserver.logSys = logSys
+
+server.DEF_LOGTARGET = "/dev/null"
+
+def _test_output(*args):
+	logSys.info(args[0])
+fail2bancmdline.output = \
+fail2banclient.output = \
+fail2banserver.output = \
+protocol.output = _test_output
+
+def _test_exit(code=0):
+	logSys.debug("Exit with code %s", code)
+	if code == 0:
+		raise ExitException()
+	else:
+		raise FailExitException()   
+fail2bancmdline.exit = \
```

Comment:
note that mocking it this way would then 1. be permanent as soon as this module gets imported, 2. would be dependent on order of tests imports.  That is why in general better to mock specifically in setup/teardown methods.  In general we might eventually start using mock module (standard since 3.4 iirc) -- very powerful and nice